keel-artifact/src/main/kotlin/com/netflix/spinnaker/keel/artifact/DockerArtifactPublisher.kt

```diff
@@ -0,0 +1,99 @@
+package com.netflix.spinnaker.keel.artifact
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.artifacts.BuildMetadata
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.artifacts.GitMetadata
+import com.netflix.spinnaker.keel.api.artifacts.PublishedArtifact
+import com.netflix.spinnaker.keel.api.artifacts.TagVersionStrategy.BRANCH_JOB_COMMIT_BY_JOB
+import com.netflix.spinnaker.keel.api.artifacts.TagVersionStrategy.SEMVER_JOB_COMMIT_BY_JOB
+import com.netflix.spinnaker.keel.api.artifacts.TagVersionStrategy.SEMVER_JOB_COMMIT_BY_SEMVER
+import com.netflix.spinnaker.keel.api.artifacts.VersioningStrategy
+import com.netflix.spinnaker.keel.api.plugins.ArtifactPublisher
+import com.netflix.spinnaker.keel.api.plugins.SupportedArtifact
+import com.netflix.spinnaker.keel.api.plugins.SupportedVersioningStrategy
+import com.netflix.spinnaker.keel.api.support.EventPublisher
+import com.netflix.spinnaker.keel.artifacts.DOCKER
+import com.netflix.spinnaker.keel.artifacts.DockerArtifact
+import com.netflix.spinnaker.keel.artifacts.DockerVersioningStrategy
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
+import org.springframework.stereotype.Component
+
+/**
+ * Built-in keel implementation of [ArtifactPublisher] that does not itself receive/retrieve artifact information
+ * but is used by keel's `POST /artifacts/events` API to notify the core of new Docker artifacts.
+ */
+@Component
+class DockerArtifactPublisher(
+  override val eventPublisher: EventPublisher,
+  private val cloudDriverService: CloudDriverService
+) : ArtifactPublisher<DockerArtifact> {
+  override val supportedArtifact = SupportedArtifact("docker", DockerArtifact::class.java)
+
+  override val supportedVersioningStrategies: List<SupportedVersioningStrategy<*>>
+    get() = listOf(
+      SupportedVersioningStrategy("docker", DockerVersioningStrategy::class.java)
+    )
+
+  override suspend fun getLatestArtifact(deliveryConfig: DeliveryConfig, artifact: DeliveryArtifact): PublishedArtifact? {
+    if (artifact !is DockerArtifact) {
+      throw IllegalArgumentException("Only Docker artifacts are supported by this implementation.")
+    }
+
+    val latestTag = cloudDriverService
+      .findDockerTagsForImage("*", artifact.name, deliveryConfig.serviceAccount)
+      .distinct()
+      .sortedWith(artifact.versioningStrategy.comparator)
+      .firstOrNull()
+
+    return if (latestTag != null) {
+      // TODO: do we need to look in a specific account? if so, then DockerArtifact should have an
```

Comment:
we're looking in all known repositories here I believe (`"*"`). I'd probably suggest changing this comment to indicate what we do?

Reply:
OK, will do. I left the `TODO` there because this was a change from the previous implementation (we were just getting the version before), so I wasn't sure what needed to go here, but followed the example of the other call to find versions/tags. Hopefully it works OK.